### PR TITLE
Type hints and doc strings: conda.cli.main_info

### PR DESCRIPTION
**Pull request by samhaese**
### Description

This PR is part of an effort to add type hints and doc strings (#13437 ).  This PR only modified conda.cli.main_info by adding type hints and rst-style doc strings, with one exception (changed the name of the argument for `dump_record` from `pkg` to `prec`).

This is my first PR for this effort, so there were a few things that I'm not sure about.  For example, how to check API documentation rendering as suggested in the issue.   Advice and feedback appreciated :)

### Miscellaneous notes

Going through the code for type hints and descriptions, I came across a few things that might deserve attention.

- Should the `get_user_site` function be updated with platformdirs?  Right now it returns possibly misleading information.
- Several of the `conda info` usages have no --json support

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:143]:
```suggestion
IGNORE_FIELDS: set[str] = {"files", "auth", "preferred_env", "priority"}
```

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:146]:
```suggestion
SKIP_FIELDS: set[str] = {
    *IGNORE_FIELDS,
    "name",
```

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:158]:
```suggestion
def dump_record(prec: PackageRecord) -> dict[str, Any]:
```

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:236]:
```suggestion
def get_info_dict(system: bool = False) -> dict[str, Any]:
```

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:365]:
```suggestion
def get_env_vars_str(info_dict: dict[str, Any]) -> str:
```

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:386]:
```suggestion
def get_main_info_str(info_dict: dict[str, Any]) -> str:
```

**Review by kenodegard — CHANGES_REQUESTED** [review]
This is a great start! Only suggestion I have is to make the return types more specific.

Regarding reviewing the rendered docs, these are auto generated as part of PR checks/GH Actions (see the ReadTheDocs check below).

File: conda/cli/main_info.py
Comment (samhaese) [reply conda/cli/main_info.py]: Of course, how could I forget!  Updated.

**Comment by samhaese** [comment]
pre-commit.ci autofix

**Comment by travishathaway** [comment]
@samhaese,

Just FYI, we build a documentation preview for each pull request:

Bildschirmfoto 2023-12-19 um 06 59 06

Click the "details" link to see it.

File: conda/cli/main_info.py
Comment (samhaese) [reply conda/cli/main_info.py]: Updated

File: conda/cli/main_info.py
Comment (samhaese) [reply conda/cli/main_info.py]: Updated

File: conda/cli/main_info.py
Comment (samhaese) [reply conda/cli/main_info.py]: Updated

File: conda/cli/main_info.py
Comment (samhaese) [reply conda/cli/main_info.py]: Updated

**Review by travishathaway — DISMISSED** [review]
I think this looks good. Waiting for @kenodegard to approve too before merging.

Thank you!

**Comment by kenodegard** [comment]
blocked by unrelated test failures, waiting on https://github.com/conda/conda/pull/13449

File: news/13437-add-type-hints-main_info
Comment (kenodegard) [inline news/13437-add-type-hints-main_info:15]:
```suggestion
* Added type hints and doc strings to `conda.cli.main_info.py`. (#13445)
```

**Review by kenodegard — APPROVED** [review]

**Review by jezdez — CHANGES_REQUESTED** [review]
Specifying the return type again in the docstring is a duplication of information and not needed because sphinx-autoapi supports autodoc's ability to [automatically derive it from the object's typehints](https://sphinx-autoapi.readthedocs.io/en/latest/how_to.html#how-to-include-type-annotations-as-types-in-rendered-docstrings)!

Thank you for updating the docstrings with blurbs of course and adding the typehints, but could you please remove the `:returns:` statements?

**Comment by kenodegard** [comment]
I think we still want the `:returns:`, it's the `:rtype:` which we can forego with sphinx-autoapi

**Comment by samhaese** [comment]
> Specifying the return type again in the docstring is a duplication of information and not needed because sphinx-autoapi supports autodoc's ability to [automatically derive it from the object's typehints](https://sphinx-autoapi.readthedocs.io/en/latest/how_to.html#how-to-include-type-annotations-as-types-in-rendered-docstrings)!
> 
> Thank you for updating the docstrings with blurbs of course and adding the typehints, but could you please remove the `:returns:` statements?

Please correct me if I'm wrong but I believe the `:returns:` directive only supplies descriptive language for sphinx autodoc, not the return type.  Do you still recommend removal?

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:26]: Let's move this into a type-checking block:

```suggestion
from typing import Any, Iterable, TYPE_CHECKING

from ..deprecations import deprecated

if TYPE_CHECKING:
    from ..models.records import PackageRecord
```

File: news/13437-add-type-hints-main_info
Comment (kenodegard) [inline news/13437-add-type-hints-main_info:15]:
```suggestion
* Added type hints and doc strings to `conda.cli.main_info`. (#13445)
```

**Review by kenodegard — CHANGES_REQUESTED** [review]

File: conda/cli/main_info.py
Comment (kenodegard) [inline conda/cli/main_info.py:243]: Since it's unused we want to deprecate it:

```suggestion
@deprecated.argument("24.9", "25.3", "system")
def get_info_dict() -> dict[str, Any]:
    """
    Returns a dictionary of contextual information.

```

**Comment by kenodegard** [comment]
> > Specifying the return type again in the docstring is a duplication of information and not needed because sphinx-autoapi supports autodoc's ability to [automatically derive it from the object's typehints](https://sphinx-autoapi.readthedocs.io/en/latest/how_to.html#how-to-include-type-annotations-as-types-in-rendered-docstrings)!
> > Thank you for updating the docstrings with blurbs of course and adding the typehints, but could you please remove the `:returns:` statements?
> 
> Please correct me if I'm wrong but I believe the `:returns:` directive only supplies descriptive language for sphinx autodoc, not the return type. Do you still recommend removal?

@samhaese you are correct, no need to remove `:returns:`

**Comment by kenodegard** [comment]
pre-commit.ci autofix

**Review by kenodegard — APPROVED** [review]

**Review by jezdez — APPROVED** [review]

**Pull request merged**
